File: backend/msfuncionario/src/main/java/com/ms/funcionario/services/FuncionarioService.java
package com.ms.funcionario.services;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.ms.funcionario.dto.CadastroDTO;
import com.ms.funcionario.dto.FuncionarioDTO;
import com.ms.funcionario.model.Funcionario;
import com.ms.funcionario.repository.FuncionarioRepository;
import org.springframework.transaction.annotation.Transactional;
import java.util.List;
import java.util.stream.Collectors;
import org.modelmapper.ModelMapper;
import org.springframework.amqp.AmqpException;
import org.springframework.amqp.rabbit.core.RabbitTemplate;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

@Service
public class FuncionarioService {

    @Autowired
    private FuncionarioRepository repository;

    @Autowired
    private ModelMapper mapper;

    @Autowired
    private RabbitTemplate rabbitTemplate;

    @Transactional
    public FuncionarioDTO saveFuncionario(FuncionarioDTO funcionarioDTO) {
        boolean cpfExistente = repository.findByCpf(funcionarioDTO.getCpf()).isPresent();
        boolean emailExistente = repository.findByEmail(funcionarioDTO.getEmail()).isPresent();

        if (cpfExistente || emailExistente) {
            throw new IllegalArgumentException("CPF ou E-mail já cadastrados.");
        }

        Funcionario funcionario = mapper.map(funcionarioDTO, Funcionario.class);

        if (funcionario.getAtivo() == null) {
            funcionario.setAtivo(true); 
        }

        Funcionario funcionarioSalvo = repository.save(funcionario);

        CadastroDTO cadastroDTO = new CadastroDTO();
        cadastroDTO.setNome(funcionarioSalvo.getNome());
        cadastroDTO.setEmail(funcionarioSalvo.getEmail());
        cadastroDTO.setPerfil("FUNCIONARIO");

        try {
            String mensagem = new ObjectMapper().writeValueAsString(cadastroDTO);

            rabbitTemplate.convertAndSend("autocadastro", "auth", mensagem);

            System.out.println("foi");
        } catch (JsonProcessingException e) {
            throw new RuntimeException("Erro ao serializar DTO para JSON ou enviar para RabbitMQ", e);
        } catch (AmqpException e) {
            throw new RuntimeException("Erro ao enviar mensagem para RabbitMQ", e);
        }

        return mapper.map(funcionarioSalvo, FuncionarioDTO.class);
    }

    @Transactional
    public FuncionarioDTO findByIdAtivo(Long id) {
        Funcionario funcionario = repository.findById(id).filter(Funcionario::getAtivo).orElseThrow(() -> new RuntimeException("Funcionário não encontrado com o ID: " + id));

        return mapper.map(funcionario, FuncionarioDTO.class);
    }

    @Transactional
    public Funcionario findById(Long id) {
        Funcionario funcionario = repository.findById(id).orElseThrow(() -> new RuntimeException("Funcionário não encontrado com o ID: " + id));

        if (!funcionario.getAtivo()) {
            throw new RuntimeException("Não é possível atualizar um funcionário inativo");
        }

        return funcionario;
    }

    @Transactional
    public List<FuncionarioDTO> listarFuncionarios() {
        List<Funcionario> ativos = repository.findAllByAtivoTrueOrderByNomeAsc();

        return ativos.stream().map(func -> mapper.map(func, FuncionarioDTO.class)).collect(Collectors.toList());
    }

    @Transactional
    public FuncionarioDTO findByEmail(String email) {
        Funcionario funcionario = repository.findByEmail(email).orElseThrow(() -> new RuntimeException("Funcionário não encontrado com o e-mail: " + email));

        return mapper.map(funcionario, FuncionarioDTO.class);
    }

    @Transactional
    public FuncionarioDTO updateFuncionario(Long id, FuncionarioDTO funcionarioDTO) {
        Funcionario funcionarioExistente = repository.findById(id).orElseThrow(() -> new RuntimeException("Funcionário não encontrado com o ID: " + id));

        funcionarioExistente.setNome(funcionarioDTO.getNome());
        funcionarioExistente.setEmail(funcionarioDTO.getEmail());
        funcionarioExistente.setTelefone(funcionarioDTO.getTelefone());

        try {
            return mapper.map(repository.save(funcionarioExistente), FuncionarioDTO.class);
        } catch (Exception e) {
            throw new RuntimeException("Erro ao atualizar funcionário: " + e.getMessage(), e);
        }
    }

    @Transactional
    public FuncionarioDTO deleteFuncionario(Long id) {
        Funcionario funcionario = repository.findById(id).orElseThrow(() -> new RuntimeException("Funcionário não encontrado com o ID: " + id));

        funcionario.setAtivo(false);

        Funcionario atualizado = repository.save(funcionario);

        return mapper.map(atualizado, FuncionarioDTO.class);
    }
}

File: backend/funcionario-service/src/main/kotlin/br/ufpr/dac/funcionario_service/resource/FuncionarioService.kt
package br.ufpr.dac.funcionario_service.resource

import br.ufpr.dac.funcionario_service.repository.IFuncionarioRepository
import br.ufpr.dac.funcionario_service.resource.dto.FuncionarioMapper
import org.springframework.amqp.core.DirectExchange
import org.springframework.amqp.rabbit.core.RabbitTemplate
import org.springframework.beans.factory.annotation.Qualifier
import utils.dto.FuncionarioOutputDTO
import org.springframework.stereotype.Service
import utils.dto.FuncionarioInputDTO
import utils.exceptions.ResourceNotFoundException
import utils.exceptions.ResourcesConflictException

@Service
class FuncionarioService(
    private val repository: IFuncionarioRepository,
    private val template: RabbitTemplate,
    @Qualifier("deactivateFuncionario") val exchange: DirectExchange
) {

    fun getAllFuncionarios(): List<FuncionarioOutputDTO> {
        return repository.findByAtivoTrue().map { FuncionarioMapper.toDTO(it) }
    }

    fun updateFuncionario(codigo: Long, funcionarioDTO: FuncionarioInputDTO): FuncionarioOutputDTO {
        val funcionario = repository.findById(codigo)
            .orElseThrow { ResourceNotFoundException("Funcionário não encontrado com o ID: ${funcionarioDTO.codigo}") }
        funcionario.nome = funcionarioDTO.nome
        funcionario.email = funcionarioDTO.email
        funcionario.telefone = funcionarioDTO.telefone
        funcionario.ativo = funcionarioDTO.ativo ?: funcionario.ativo

        return FuncionarioMapper.toDTO(repository.save(funcionario))
    }

    fun saveFuncionario(funcionario: FuncionarioInputDTO): FuncionarioOutputDTO {
        try {
            val registry = repository.save(FuncionarioMapper.toDomain(funcionario))
            return FuncionarioMapper.toDTO(registry)
        } catch (ex: Exception) {
            throw ResourcesConflictException("Usuário duplicado")
        }
    }

    fun deactivateFuncionario(codigo: Long): FuncionarioOutputDTO {
        repository.findByCodigoAndAtivoTrue(codigo)?.let { funcionario ->
            funcionario.ativo = false
            template.convertAndSend(exchange.name, "funcionario", codigo.toString())
            return FuncionarioMapper.toDTO(repository.save(funcionario))
        }

        throw ResourceNotFoundException("Funcionário não encontrado com o ID: $codigo")
    }

    fun getFuncionarioById(codigo: Long): FuncionarioOutputDTO {
        repository.findByCodigoAndAtivoTrue(codigo)?.let { funcionario ->
            return FuncionarioMapper.toDTO(funcionario)
        }

        throw ResourceNotFoundException("Funcionário não encontrado com o ID: $codigo")
    }
}
